File: samples/python-payments-service/app/payments/processor.py
"""Payment processing — authenticates then records a payment."""
from app.auth.login import authenticate


class PaymentError(Exception):
    """Raised when a payment is rejected."""


def process_payment(username, password, amount, conn=None):
    """Authenticate the user and record a captured payment. Returns payment id."""
    from app.db import get_connection

    conn = conn or get_connection()
    user_id = authenticate(username, password, conn=conn)
    if amount <= 0:
        raise PaymentError("amount must be positive")
    cur = conn.execute(
        "INSERT INTO payments (user_id, amount, status) VALUES (?, ?, ?)",
        (user_id, amount, "captured"),
    )
    conn.commit()
    return cur.lastrowid

File: samples/node-catalog-service/src/catalog/processor.js
"use strict";

// Catalog processing — authenticates then records a catalog item.
const { authenticate } = require("../auth/login");

class CatalogError extends Error {}

function addItem(username, password, item, conn) {
  authenticate(username, password, conn);
  if (!item || !item.name) {
    throw new CatalogError("item name is required");
  }
  const record = { id: conn.catalog.length + 1, name: item.name, price: item.price || 0 };
  conn.catalog.push(record);
  return record;
}

module.exports = { CatalogError, addItem };
